ls1intum/Artemis
# Fix for Exercise Participation Usability Issues (Modeling, Text, File Upload)

### Checklist

### Motivation and Context
This PR addresses [this](https://github.com/ls1intum/Artemis/issues/1297) issue.

### Description
1. Introduce common participation header for Modeling, Text and File Upload Exercises
2. New differentiation in submission result status between `SUBMITTED` (user has submitted and can still submit) and `SUBMITTED_WAITING_FOR_GRADING` (user has submitted and can no longer submit but has to wait for the grading result to arrive)
3. When a student submits a Modeling, Text, File upload submission, the submission status will now change immediately to reflect the updated status
4. For the Modeling Editor, an unsaved changes guard was triggered sometimes when leaving the route even though no changes were actually made. The was due to the dimensions of the canvas automatically being shrunk to the actually used up content area by the Apollon Editor. The fix is to ignore the dimension when comparing the saved model with the current model.
5. Slightly improved styling for the Text Editor (incl. nicer focus state)

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Overview
3. Participate in a Modeling, Text and File Upload Exercise
4. Check that the new participation header shows the correct information
5. Make a submission and check that the submission status updates right away (it should show "Submitted, you can still submit updates" if the due date is not reached yet)
6. For the modeling editor, make sure that the "unsaved changes" warning only appears if actual changes were made

### Screenshots
![OnPaste 20200420-130213](https://user-images.githubusercontent.com/7109225/79744854-68886380-8307-11ea-885c-fe81a1ca5d29.png)
![OnPaste 20200420-130319](https://user-images.githubusercontent.com/7109225/79744857-69b99080-8307-11ea-8e48-2c7bc30b4249.png)
![OnPaste 20200420-130357](https://user-images.githubusercontent.com/7109225/79744862-6a522700-8307-11ea-96ce-06485d7dbb87.png)

There were also some display issues with the alert not having any space around it that is now resolved:
![OnPaste 20200420-131056](https://user-images.githubusercontent.com/7109225/79745348-64107a80-8308-11ea-837a-6e41021bc7e5.png)
